README.md
# DCdumper
Tool for producing [redump compliant](http://wiki.redump.org/index.php?title=GD-Rom_Dumping_Guide) dumps of Sega Dreamcast GD-ROM discs


Changelog.txt
20110919 (0.42a)
+FIX:Read errors apparent since 0.3a
+FIX:Writing hashes to file

20110918 (0.41a)
+FIX:Suppress some strings that shouldn't be shown on fake reads

20110917 (0.4a)
More options:
 -ff -ft : Fake reads to try and kick program out of a rut in the event of read errors (default true)
 -r#     : Retry attempts per pass in event of read errors (min 0, max 20, default 3)
 -sf -st : Step back. Fake reads from end to start after a pass. Perhaps gentler than jumping off a cliff (default true)
 
+Added retry attempts per pass in event of read errors (-r)
+Added fake reads to try and kick program out of a rut in the event of read errors (-f)
+Added step back. Fake reads from end to start, perhaps gentler than jumping off a cliff (-s)

20110917 (0.3a)
+No longer use akrip32, use DeviceIO (SPTD)
+No longer use startstop.exe, use DeviceIO (Load device)
+Section hashes stored to file as successful reads are completed (better resume function)
+Log file created (mirrors console)
+Pass drive letter, not scsi id

20110915 (0.2a)
+Options no longer hard-coded
 -h      : Basic help
 -c#     : Set cut-off sector, after which the disc will be dumped in one big lump (min 44990 (dump in 1 lump), max 549150 (impose no cut off), default 549150)
 -dt -df : Set whether or not to delete temporary files when creating dense.bin
 -i#     : Set number of sectors per section (min 26, max 504161, default 10289)
 -p#     : Set number of passes (min 2, max 100, default 10)

20110914 (0.1a)
Initial release
+Dumps in 10289-sector sections
+Makes up to 10 passes
-Needs both scsi device id (#:#:#) and drive letter passed
-Not entirely safe (regarding passing 2 drive definitions)
-Options hard-coded


src/md5.h
/* MD5
  converted to C++ class by Frank Thilo (thilo@unix-ag.org)
  for bzflag (http://www.bzflag.org)

    based on:

    md5.h and md5.c
    reference implementation of RFC 1321

    Copyright (C) 1991-2, RSA Data Security, Inc. Created 1991. All
 rights reserved.

 License to copy and use this software is granted provided that it
 is identified as the "RSA Data Security, Inc. MD5 Message-Digest
 Algorithm" in all material mentioning or referencing this software
 or this function.

 License is also granted to make and use derivative works provided
 that such works are identified as "derived from the RSA Data
 Security, Inc. MD5 Message-Digest Algorithm" in all material
 mentioning or referencing the derived work.

 RSA Data Security, Inc. makes no representations concerning either
 the merchantability of this software or the suitability of this
 software for any particular purpose. It is provided "as is"
 without express or implied warranty of any kind.

 These notices must be retained in any copies of any part of this
 documentation and/or software.

 */

#ifndef BZF_MD5_H
#define BZF_MD5_H

#include <string>
#include <iostream>


// a small class for calculating MD5 hashes of strings or byte arrays
// it is not meant to be fast or secure
//
// usage: 1) feed it blocks of uchars with update()
//      2) finalize()
//      3) get hexdigest() string
//      or
//      MD5(std::string).hexdigest()
//
// assumes that char is 8 bit and int is 32 bit
class MD5
{
public:
   typedef unsigned int size_type; // must be 32bit

   MD5();
   MD5(const std::string& text);
   void update(const unsigned char *buf, size_type length);
   void update(const char *buf, size_type length);
   MD5& finalize();
   std::string hexdigest() const;
   friend std::ostream& operator<<(std::ostream&, MD5 md5);

private:
   void init();
   typedef unsigned char uint1; //  8bit
   typedef unsigned int uint4;  // 32bit
   enum {blocksize = 64}; // VC6 won't eat a const static int here

   void transform(const uint1 block[blocksize]);
   static void decode(uint4 output[], const uint1 input[], size_type len);
   static void encode(uint1 output[], const uint4 input[], size_type len);

   bool finalized;
   uint1 buffer[blocksize]; // bytes that didn't fit in last 64 byte chunk
   uint4 count[2];   // 64bit counter for number of bits (lo, hi)
   uint4 state[4];   // digest so far
   uint1 digest[16]; // the result

   // low level logic operations
   static inline uint4 F(uint4 x, uint4 y, uint4 z);
   static inline uint4 G(uint4 x, uint4 y, uint4 z);
   static inline uint4 H(uint4 x, uint4 y, uint4 z);
   static inline uint4 I(uint4 x, uint4 y, uint4 z);
   static inline uint4 rotate_left(uint4 x, int n);
   static inline void FF(uint4 &a, uint4 b, uint4 c, uint4 d, uint4 x, uint4 s, uint4 ac);
   static inline void GG(uint4 &a, uint4 b, uint4 c, uint4 d, uint4 x, uint4 s, uint4 ac);
   static inline void HH(uint4 &a, uint4 b, uint4 c, uint4 d, uint4 x, uint4 s, uint4 ac);
   static inline void II(uint4 &a, uint4 b, uint4 c, uint4 d, uint4 x, uint4 s, uint4 ac);
};

 std::string md5(const std::string str);

#endif

src/md5.cpp
/* MD5
  converted to C++ class by Frank Thilo (thilo@unix-ag.org)
  for bzflag (http://www.bzflag.org)

    based on:

    md5.h and md5.c
    reference implemantion of RFC 1321

    Copyright (C) 1991-2, RSA Data Security, Inc. Created 1991. All
 rights reserved.

 License to copy and use this software is granted provided that it
 is identified as the "RSA Data Security, Inc. MD5 Message-Digest
 Algorithm" in all material mentioning or referencing this software
 or this function.

 License is also granted to make and use derivative works provided
 that such works are identified as "derived from the RSA Data
 Security, Inc. MD5 Message-Digest Algorithm" in all material
 mentioning or referencing the derived work.

 RSA Data Security, Inc. makes no representations concerning either
 the merchantability of this software or the suitability of this
 software for any particular purpose. It is provided "as is"
 without express or implied warranty of any kind.

 These notices must be retained in any copies of any part of this
 documentation and/or software.

 */

/* interface header */
#include "md5.h"

/* system implementation headers */
#include <stdio.h>


// Constants for MD5Transform routine.
#define S11 7
#define S12 12
#define S13 17
#define S14 22
#define S21 5
#define S22 9
#define S23 14
#define S24 20
#define S31 4
#define S32 11
#define S33 16
#define S34 23
#define S41 6
#define S42 10
#define S43 15
#define S44 21

///////////////////////////////////////////////

// F, G, H and I are basic MD5 functions.
inline MD5::uint4 MD5::F(uint4 x, uint4 y, uint4 z) {
   return x&y | ~x&z;
}

inline MD5::uint4 MD5::G(uint4 x, uint4 y, uint4 z) {
   return x&z | y&~z;
}

inline MD5::uint4 MD5::H(uint4 x, uint4 y, uint4 z) {
   return x^y^z;
}

inline MD5::uint4 MD5::I(uint4 x, uint4 y, uint4 z) {
   return y ^ (x | ~z);
}

// rotate_left rotates x left n bits.
inline MD5::uint4 MD5::rotate_left(uint4 x, int n) {
   return (x << n) | (x >> (32-n));
}

// FF, GG, HH, and II transformations for rounds 1, 2, 3, and 4.
// Rotation is separate from addition to prevent recomputation.
inline void MD5::FF(uint4 &a, uint4 b, uint4 c, uint4 d, uint4 x, uint4 s, uint4 ac) {
   a = rotate_left(a+ F(b,c,d) + x + ac, s) + b;
}

inline void MD5::GG(uint4 &a, uint4 b, uint4 c, uint4 d, uint4 x, uint4 s, uint4 ac) {
   a = rotate_left(a + G(b,c,d) + x + ac, s) + b;
}

inline void MD5::HH(uint4 &a, uint4 b, uint4 c, uint4 d, uint4 x, uint4 s, uint4 ac) {
   a = rotate_left(a + H(b,c,d) + x + ac, s) + b;
}

inline void MD5::II(uint4 &a, uint4 b, uint4 c, uint4 d, uint4 x, uint4 s, uint4 ac) {
   a = rotate_left(a + I(b,c,d) + x + ac, s) + b;
}

//////////////////////////////////////////////

// default ctor, just initailize
 MD5::MD5()
{
   init();
}

//////////////////////////////////////////////

// nifty shortcut ctor, compute MD5 for string and finalize it right away
 MD5::MD5(const std::string &text)
{
   init();
   update(text.c_str(), text.length());
   finalize();
}

//////////////////////////////

void MD5::init()
{
   finalized=false;

   count[0] = 0;
   count[1] = 0;

   // load magic initialization constants.
   state[0] = 0x67452301;
   state[1] = 0xefcdab89;
   state[2] = 0x98badcfe;
   state[3] = 0x10325476;
}

//////////////////////////////

// decodes input (unsigned char) into output (uint4). Assumes len is a multiple of 4.
void MD5::decode(uint4 output[], const uint1 input[], size_type len)
{
   for (unsigned int i = 0, j = 0; j < len; i++, j += 4)
     output[i] = ((uint4)input[j]) | (((uint4)input[j+1]) << 8) |
       (((uint4)input[j+2]) << 16) | (((uint4)input[j+3]) << 24);
}

//////////////////////////////

// encodes input (uint4) into output (unsigned char). Assumes len is
// a multiple of 4.
void MD5::encode(uint1 output[], const uint4 input[], size_type len)
{
   for (size_type i = 0, j = 0; j < len; i++, j += 4) {
     output[j] = input[i] & 0xff;
     output[j+1] = (input[i] >> 8) & 0xff;
     output[j+2] = (input[i] >> 16) & 0xff;
     output[j+3] = (input[i] >> 24) & 0xff;
   }
}

//////////////////////////////

// apply MD5 algo on a block
void MD5::transform(const uint1 block[blocksize])
{
   uint4 a = state[0], b = state[1], c = state[2], d = state[3], x[16];
   decode (x, block, blocksize);

   /* Round 1 */
   FF (a, b, c, d, x[ 0], S11, 0xd76aa478); /* 1 */
   FF (d, a, b, c, x[ 1], S12, 0xe8c7b756); /* 2 */
   FF (c, d, a, b, x[ 2], S13, 0x242070db); /* 3 */
   FF (b, c, d, a, x[ 3], S14, 0xc1bdceee); /* 4 */
   FF (a, b, c, d, x[ 4], S11, 0xf57c0faf); /* 5 */
   FF (d, a, b, c, x[ 5], S12, 0x4787c62a); /* 6 */
   FF (c, d, a, b, x[ 6], S13, 0xa8304613); /* 7 */
   FF (b, c, d, a, x[ 7], S14, 0xfd469501); /* 8 */
   FF (a, b, c, d, x[ 8], S11, 0x698098d8); /* 9 */
   FF (d, a, b, c, x[ 9], S12, 0x8b44f7af); /* 10 */
   FF (c, d, a, b, x[10], S13, 0xffff5bb1); /* 11 */
   FF (b, c, d, a, x[11], S14, 0x895cd7be); /* 12 */
   FF (a, b, c, d, x[12], S11, 0x6b901122); /* 13 */
   FF (d, a, b, c, x[13], S12, 0xfd987193); /* 14 */
   FF (c, d, a, b, x[14], S13, 0xa679438e); /* 15 */
   FF (b, c, d, a, x[15], S14, 0x49b40821); /* 16 */

   /* Round 2 */
   GG (a, b, c, d, x[ 1], S21, 0xf61e2562); /* 17 */
   GG (d, a, b, c, x[ 6], S22, 0xc040b340); /* 18 */
   GG (c, d, a, b, x[11], S23, 0x265e5a51); /* 19 */
   GG (b, c, d, a, x[ 0], S24, 0xe9b6c7aa); /* 20 */
   GG (a, b, c, d, x[ 5], S21, 0xd62f105d); /* 21 */
   GG (d, a, b, c, x[10], S22,  0x2441453); /* 22 */
   GG (c, d, a, b, x[15], S23, 0xd8a1e681); /* 23 */
   GG (b, c, d, a, x[ 4], S24, 0xe7d3fbc8); /* 24 */
   GG (a, b, c, d, x[ 9], S21, 0x21e1cde6); /* 25 */
   GG (d, a, b, c, x[14], S22, 0xc33707d6); /* 26 */
   GG (c, d, a, b, x[ 3], S23, 0xf4d50d87); /* 27 */
   GG (b, c, d, a, x[ 8], S24, 0x455a14ed); /* 28 */
   GG (a, b, c, d, x[13], S21, 0xa9e3e905); /* 29 */
   GG (d, a, b, c, x[ 2], S22, 0xfcefa3f8); /* 30 */
   GG (c, d, a, b, x[ 7], S23, 0x676f02d9); /* 31 */
   GG (b, c, d, a, x[12], S24, 0x8d2a4c8a); /* 32 */

   /* Round 3 */
   HH (a, b, c, d, x[ 5], S31, 0xfffa3942); /* 33 */
   HH (d, a, b, c, x[ 8], S32, 0x8771f681); /* 34 */
   HH (c, d, a, b, x[11], S33, 0x6d9d6122); /* 35 */
   HH (b, c, d, a, x[14], S34, 0xfde5380c); /* 36 */
   HH (a, b, c, d, x[ 1], S31, 0xa4beea44); /* 37 */
   HH (d, a, b, c, x[ 4], S32, 0x4bdecfa9); /* 38 */
   HH (c, d, a, b, x[ 7], S33, 0xf6bb4b60); /* 39 */
   HH (b, c, d, a, x[10], S34, 0xbebfbc70); /* 40 */
   HH (a, b, c, d, x[13], S31, 0x289b7ec6); /* 41 */
   HH (d, a, b, c, x[ 0], S32, 0xeaa127fa); /* 42 */
   HH (c, d, a, b, x[ 3], S33, 0xd4ef3085); /* 43 */
   HH (b, c, d, a, x[ 6], S34,  0x4881d05); /* 44 */
   HH (a, b, c, d, x[ 9], S31, 0xd9d4d039); /* 45 */
   HH (d, a, b, c, x[12], S32, 0xe6db99e5); /* 46 */
   HH (c, d, a, b, x[15], S33, 0x1fa27cf8); /* 47 */
   HH (b, c, d, a, x[ 2], S34, 0xc4ac5665); /* 48 */

   /* Round 4 */
   II (a, b, c, d, x[ 0], S41, 0xf4292244); /* 49 */
   II (d, a, b, c, x[ 7], S42, 0x432aff97); /* 50 */
   II (c, d, a, b, x[14], S43, 0xab9423a7); /* 51 */
   II (b, c, d, a, x[ 5], S44, 0xfc93a039); /* 52 */
   II (a, b, c, d, x[12], S41, 0x655b59c3); /* 53 */
   II (d, a, b, c, x[ 3], S42, 0x8f0ccc92); /* 54 */
   II (c, d, a, b, x[10], S43, 0xffeff47d); /* 55 */
   II (b, c, d, a, x[ 1], S44, 0x85845dd1); /* 56 */
   II (a, b, c, d, x[ 8], S41, 0x6fa87e4f); /* 57 */
   II (d, a, b, c, x[15], S42, 0xfe2ce6e0); /* 58 */
   II (c, d, a, b, x[ 6], S43, 0xa3014314); /* 59 */
   II (b, c, d, a, x[13], S44, 0x4e0811a1); /* 60 */
   II (a, b, c, d, x[ 4], S41, 0xf7537e82); /* 61 */
   II (d, a, b, c, x[11], S42, 0xbd3af235); /* 62 */
   II (c, d, a, b, x[ 2], S43, 0x2ad7d2bb); /* 63 */
   II (b, c, d, a, x[ 9], S44, 0xeb86d391); /* 64 */

   state[0] += a;
   state[1] += b;
   state[2] += c;
   state[3] += d;

   // Zeroize sensitive information.
   memset(x, 0, sizeof x);
}

//////////////////////////////

// MD5 block update operation. Continues an MD5 message-digest
// operation, processing another message block
void MD5::update(const unsigned char input[], size_type length)
{
   // compute number of bytes mod 64
   size_type index = count[0] / 8 % blocksize;

   // Update number of bits
   if ((count[0] += (length << 3)) < (length << 3))
     count[1]++;
   count[1] += (length >> 29);

   // number of bytes we need to fill in buffer
   size_type firstpart = 64 - index;

   size_type i;

   // transform as many times as possible.
   if (length >= firstpart)
   {
     // fill buffer first, transform
     memcpy(&buffer[index], input, firstpart);
     transform(buffer);

     // transform chunks of blocksize (64 bytes)
     for (i = firstpart; i + blocksize <= length; i += blocksize)
       transform(&input[i]);

     index = 0;
   }
   else
     i = 0;

   // buffer remaining input
   memcpy(&buffer[index], &input[i], length-i);
}

//////////////////////////////

// for convenience provide a verson with signed char
void MD5::update(const char input[], size_type length)
{
   update((const unsigned char*)input, length);
}

//////////////////////////////

// MD5 finalization. Ends an MD5 message-digest operation, writing the
// the message digest and zeroizing the context.
 MD5& MD5::finalize()
{
   static unsigned char padding[64] = {
     0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
     0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
     0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
   };

   if (!finalized) {
     // Save number of bits
     unsigned char bits[8];
     encode(bits, count, 8);

     // pad out to 56 mod 64.
     size_type index = count[0] / 8 % 64;
     size_type padLen = (index < 56) ? (56 - index) : (120 - index);
     update(padding, padLen);

     // Append length (before padding)
     update(bits, 8);

     // Store state in digest
     encode(digest, state, 16);

     // Zeroize sensitive information.
     memset(buffer, 0, sizeof buffer);
     memset(count, 0, sizeof count);

     finalized=true;
   }

   return *this;
}

//////////////////////////////

// return hex representation of digest as string
 std::string MD5::hexdigest() const
{
   if (!finalized)
   {
	 printf("NOT FINALISED IDIOT");
     return "";
   }

   char buf[33];
   for (int i=0; i<16; i++)
     sprintf(buf+i*2, "%02x", digest[i]);
   buf[32]=0;

   return std::string(buf);
}

//////////////////////////////

 std::ostream& operator<<(std::ostream& out, MD5 md5)
{
   return out << md5.hexdigest();
}

//////////////////////////////

 std::string md5(const std::string str)
{
     MD5 md5 = MD5(str);

     return md5.hexdigest();
}

src/SectorMap.cpp
#ifndef SECTOR_MAP
#define SECTOR_MAP

#include "Section.cpp"

extern char g_driveletter;
extern int g_special, g_passes, g_interval, g_cutoff, g_retries;
extern bool g_deltmp, g_fakereads, g_stepback;
extern ofstream g_log;

class SectorMap
{
public:
  Section* sections;
  int numsections;

  SectorMap()
  {
	if(g_cutoff==44990)
	{
	  sections=new Section[1];
	  numsections=1;
	  sections[0]=Section(44990,504161);
	}
	else
	{
	  int startsector=44990;
	  int ceiling=g_cutoff-44990+1;
	  int numwhole=ceiling/g_interval;
	  bool partial=ceiling%g_interval!=0||g_cutoff!=549150;
	  sections=new Section[numwhole+partial];
	  numsections=numwhole+partial;
	  for(int i=0;i<numwhole;i++)
	  {
	    sections[i]=Section(startsector,g_interval);
	    startsector+=g_interval;
	  }
	  if(partial)
	  {
	    sections[numsections-1]=Section(startsector,549151-startsector);
	  }
	}
  }

  void pass()
  {
	if(!isdone())
	{
	  for(int i=0;i<numsections;i++)
	  {
		if(!sections[i].done)
		{
	      printandlogtext("Reading section ",16);
		  printandlognumber(i+1);
		  printandlogtext(": ",2);
		  printandlogtext(sections[i].name.substr(0,13));
		  sections[i].read();
	    }
	  }
	  if(g_stepback)
	  {
		Section fake;
		for(int i=549000;i>44990;i-=20000)
		{
		  fake=Section(i,26);
		  fake.read(g_retries,true);
		  fake.deletefiles();
		}
	  }
	}
  }

  bool isdone()
  {
	for(int i=0;i<numsections;i++)
	{
	  if(!sections[i].done)
	  {
	    return false;
	  }
	}
	return true;
  }

};

#endif


src/main.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string>
#include <iostream>
#include "SectorMap.cpp"
#include "md5.h"

using namespace std;

int parseargs(int argc, char* argv[]);

//g_special unused, possible todo, auto settings for certain drives
static char g_driveletter;
static int g_special=0, g_passes=10, g_interval=10289, g_cutoff=549150, g_retries=3;
static bool g_deltmp=true, g_fakereads=true, g_stepback=true;
static ofstream g_log("dcd.log",ios::app|ios::out);

int main( int argc, char *argv[] )
{

  if(parseargs(argc,argv)!=0)
	return 1;
  
  
  g_log.write("\n\n---\n",7);
  g_log.write("DCdumper.exe ",13);
  for(int i=0;i<argc;i++)
  {
	printandlogtext(string(argv[i]));
    printandlogtext(" ",1);
  }
  printandlogtext("\n\n",2);

  //checks
  HANDLE testhandle;
  if((testhandle=Section().OpenVolume(g_driveletter))==INVALID_HANDLE_VALUE)
  {
    printandlogtext("Could not get handle to CD (possible wrong input).\n",51);
	g_log.close();
    return 1;
  }
  printandlogtext("Handle acquired.\n",17);

  printandlogtext("Load disc: ",11);
  DWORD dw;
  if (!DeviceIoControl(testhandle, IOCTL_STORAGE_LOAD_MEDIA, NULL, 0, NULL, 0, &dw, NULL))
  {
    printandlogtext("Failed.\n",8);
	g_log.close();
	return 1;
  }
  printandlogtext("Done.\n",6);
  Section().CloseVolume(testhandle);


  //[todo] read toc
  if (false)
  {
    printandlogtext( "Error reading TOC\n",18);
	g_log.close();
    return 1;
  }

  if(false)
  {
    printandlogtext("TOC indicates trap disc not loaded (too many tracks).\n",54);
	g_log.close();
	return 1;
  }

  //printandlogtext("TOC read.\n",10);
  //[/todo]

  SectorMap map=SectorMap();
  printandlogtext("Sector map created.\n",20);
  
  for(int i=0;i<g_passes;i++)
  {
	if(!map.isdone())
	{
	  printandlogtext("\n..................:::::::::::::::: PASS ",41);
	  printandlognumber(i+1);
	  printandlogtext(" ::::::::::::::::..................\n",36);
	  map.pass();
	}
	else
	{
	  break;
	}
  }

  if(map.isdone())
  {
	//all dumped successfully, create dense.bin
	if(map.numsections==1)
	{
	  //rename to dense.bin
	  rename(map.sections[0].name.c_str(),"dense.bin");
	  printandlogtext("Renamed to dense.bin, pass this to ice.exe\n",43);
	}
	else
	{
	  printandlogtext("Creating dense.bin: ",20);
	  ofstream outfile("dense.bin",ios::binary);
	  char* buffer=new char[2352];
	  for(int i=0;i<map.numsections;i++)
	  {
	    ifstream infile(map.sections[i].name,ios::binary);
	    for(int j=0;j<map.sections[i].numsectors;j++)
		{
		  infile.read(buffer, 2352);
		  outfile.write(buffer, 2352);
		}
		infile.close();
	  }
	  outfile.close();
	  printandlogtext("Done. Pass this to ice.exe\n",27);
	  if(g_deltmp)
	  {
	    for(int i=0;i<map.numsections;i++)
	    {
		  map.sections[i].deletefiles();
	    }
	  }

	  g_log.close();
	  return 0;
	}
  }
  else
  {
	for(int i=0;i<map.numsections;i++)
	{
	  if(!map.sections[i].done)
	  {
		printandlogtext("Sector read error between sectors ",34);
		printandlognumber(map.sections[i].initialsector);
		printandlogtext(" and ",5);
		printandlognumber(map.sections[i].initialsector+map.sections[i].numsectors-1);
		printandlogtext(" inclusive.\n",12);
	  }
	}
	printandlogtext("Clean the disc and try again.\n---",33);

	g_log.close();
	return 1;
  }
}

int parseargs(int argc, char *argv[])
{
  string helptest,currop;
  if(argc==1)
  {
    printf("Not enough arguments. -h for help.\n");
    printf("Usage: DCDump.exe #:#:# drive_letter [options]\n");
    return 1;
  }
  else if(argc>=2)
  {
	helptest=string(argv[1]);
	if(helptest==string("-h"))
	{
	  printf("Author: jamjam\n");
	  printf("Version: 0.42a\n");
	  printf("Usage: ");
	  printf(argv[0]);
	  printf(" drive_letter [options]\n\n");
	  printf("[options]\n");
	  printf("-h      - This help.\n\n");
	  printf("-c#     - Cut-off: After which disc is dumped in one lump instead of sections\n");
	  printf("          Min: 44990 (all at once)  Max: 549150 (no cut-off)  Default: 549150\n\n");
	  printf("-dt -df - Delete temporary files: Delete sections and hashes after creating dense.bin\n");
	  printf("          Default true\n\n");
	  printf("-ft -ff - Fake reads: Fake read nearby to try and kick program out of read errors\n");
	  printf("          Default true\n\n");
	  printf("-i#     - Set number of sectors per section\n");
	  printf("          Min: 26  Max: 504161 (all at once)  Default: 10289\n\n");
	  printf("-p#     - Passes: Set maximum number of passes\n");
	  printf("          Min: 2  Max: 100  Default: 10\n\n");
	  printf("-st -sf - Step back: Fake reads after a pass to step to beginning\n");
	  printf("          Default: true\n\n");
	  printf("-t#     - Reread attempts: Max number of re-read attempts on a read error\n");
	  printf("          Min: 0  Max: 20  Default: 3\n\n");
	  return 1;
	}
	else if(argv[1][0]=='-')
	{
	  printf("Core commandline arguments missing (drive letter)\n");
	  printf("Usage: DCDump.exe drive_letter [options]\n");
	  return 1;
	}
	else if((argv[1][0]<'a'||argv[1][0]>'z')&&(argv[1][0]<'A'||argv[1][0]>'Z'))
	{
	  printf("Drive letter invalid (must be in range a-z or A-Z)");
	  return 1;
	}
	else
	{
	  g_driveletter=argv[1][0];
	  for(int i=2;i<argc;i++)
	  {
	    //for every option
		currop=string(argv[i]);
		if(*(argv[i])!='-'||currop.length()<2)
		{
		  printf("\"");
		  printf(argv[i]);
		  printf("\" is not a valid option...skipping.\n");
		}
		else if(currop.at(1)=='c'||currop.at(1)=='C')
		{
		  //cut-off point
		  if(currop.length()<3)
		  {
		    printf("\"");
		    printf(argv[i]);
		    printf("\" is not a valid option...skipping.\n");
		  }
		  else
		  {
			int cutoff=atoi(currop.substr(2,currop.length()-2).c_str());
			if(cutoff>44989&&cutoff<549151)
			{
			  g_cutoff=cutoff;
			}
			else
			{
			  printf("\"");
		      printf(argv[i]);
		      printf("\" is not a valid option...skipping.\n");
			}
	      }
		}
		else if(currop.at(1)=='d'||currop.at(1)=='D')
		{
		  //delete tmp files?
		  if(currop.length()<3)
		  {
		    printf("\"");
		    printf(argv[i]);
		    printf("\" is not a valid option...skipping.\n");
		  }
		  else
		  {
			if(currop.at(2)=='f'||currop.at(2)=='F')
			{
			  g_deltmp=false;
			}
			else if(currop.at(2)=='t'||currop.at(2)=='T')
			{
			  g_deltmp=true;
			}
			else
			{
			  printf("\"");
		      printf(argv[i]);
		      printf("\" is not a valid option...skipping.\n");
			}
	      }
		}
		else if(currop.at(1)=='f'||currop.at(1)=='F')
		{
		  //try fake reads when read error?
		  if(currop.length()<3)
		  {
		    printf("\"");
		    printf(argv[i]);
		    printf("\" is not a valid option...skipping.\n");
		  }
		  else
		  {
			if(currop.at(2)=='f'||currop.at(2)=='F')
			{
			  g_fakereads=false;
			}
			else if(currop.at(2)=='t'||currop.at(2)=='T')
			{
			  g_fakereads=true;
			}
			else
			{
			  printf("\"");
		      printf(argv[i]);
		      printf("\" is not a valid option...skipping.\n");
			}
	      }
		}
		else if(currop.at(1)=='i'||currop.at(1)=='I')
		{
		  //interval
		  if(currop.length()<3)
		  {
		    printf("\"");
		    printf(argv[i]);
		    printf("\" is not a valid option...skipping.\n");
		  }
		  else
		  {
			int interval=atoi(currop.substr(2,currop.length()-2).c_str());
			if(interval>25&&interval<504162)
			{
			  g_interval=interval;
			}
			else
			{
			  printf("\"");
		      printf(argv[i]);
		      printf("\" is not a valid option...skipping.\n");
			}
	      }
		}
		else if(currop.at(1)=='p'||currop.at(1)=='P')
		{
		  //number of passes
		  if(currop.length()<3)
		  {
		    printf("\"");
		    printf(argv[i]);
		    printf("\" is not a valid option...skipping.\n");
		  }
		  else
		  {
			int passes=atoi(currop.substr(2,currop.length()-2).c_str());
			if(passes>1&&passes<101)
			{
			  g_passes=passes;
			}
			else
			{
			  printf("\"");
		      printf(argv[i]);
		      printf("\" is not a valid option...skipping.\n");
			}
	      }
		}
		else if(currop.at(1)=='s'||currop.at(1)=='S')
		{
		  //try fake reads when read error?
		  if(currop.length()<3)
		  {
		    printf("\"");
		    printf(argv[i]);
		    printf("\" is not a valid option...skipping.\n");
		  }
		  else
		  {
			if(currop.at(2)=='f'||currop.at(2)=='F')
			{
			  g_stepback=false;
			}
			else if(currop.at(2)=='t'||currop.at(2)=='T')
			{
			  g_stepback=true;
			}
			else
			{
			  printf("\"");
		      printf(argv[i]);
		      printf("\" is not a valid option...skipping.\n");
			}
	      }
		}
		else if(currop.at(1)=='t'||currop.at(1)=='T')
		{
		  //number of retries per read attempt
		  if(currop.length()<3)
		  {
		    printf("\"");
		    printf(argv[i]);
		    printf("\" is not a valid option...skipping.\n");
		  }
		  else
		  {
			int retries=atoi(currop.substr(2,currop.length()-2).c_str());
			if(retries>=0&&retries<21)
			{
			  g_retries=retries;
			}
			else
			{
			  printf("\"");
		      printf(argv[i]);
		      printf("\" is not a valid option...skipping.\n");
			}
	      }
		}
		else
		{
		  printf("\"");
		  printf(argv[i]);
		  printf("\" is not a valid option...skipping.\n");
		}
	  }
	  return 0;
	}
  }
  return 1;
}